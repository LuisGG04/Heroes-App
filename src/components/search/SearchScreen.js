import React, { useMemo } from "react";
import { useLocation, useNavigate } from "react-router-dom";
import queryString from "query-string";
import { useForm } from "../../hooks/useForm";
import { getHeroByName } from "../../selectors/getHeroByName";
import { HeroCard } from "../hero/HeroCard";

export const SearchScreen = () => {

    const navigate = useNavigate();
    const location = useLocation();

    const {q = ""} = queryString.parse(location.search);

    const [formValues, handleInputChange] = useForm({
        searchText: q
    });

    const {searchText} = formValues
    const heroFilter = useMemo(() => getHeroByName(q),[q]);

    const handleSearch = (e) => {
        e.preventDefault();
        navigate(`?q=${searchText}`);
    }

    return (
        <>
            <h1>Búsquedas</h1>
            <hr />

            <div className="row">
                <div className="col-5">
                    <h3>Buscar</h3>
                    <hr />

                    <form onSubmit={handleSearch}>

                        <input type="text" placeholder="Buscar un héroe" className="form-control" name="searchText" autoComplete="off" value={searchText} onChange={handleInputChange}/>
                        <button className="btn btn-outline-primary mt-2" type="submit">Buscar</button>

                    </form>
                </div>

                <div className="col-7">
                    <h3>Resultados</h3>
                    <hr />

                    {(q === "") ? <div className="alert alert-info">Buscar un héroe</div> : (heroFilter.length === 0) && <div className="alert alert-danger">No hay resultados: {q}</div>}

                    {heroFilter.map(hero => (
                        <HeroCard 
                            key={hero.id}
                            {...hero}
                        />
                    ))}
                </div>
            </div>

        </>
    )
}
