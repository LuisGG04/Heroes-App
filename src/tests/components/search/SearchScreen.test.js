import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";

const mockNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"), 
    useNavigate: () => mockNavigate
}));

describe("Pruebas en SearchScreen", () => {
    
    test("Debe de mostrar el componente", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search"]}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(".alert-info").text().trim()).toBe("Buscar un hÃ©roe");
    });

    test("Debe de mostrar a Batman y el input con el valor del queryString", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=batman"]}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect(wrapper.find("input").prop("value")).toBe("batman");
    });
    
    test("Debe de mostrar el mensaje 'No hay resultados' en el componente", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=batman778"]}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect(wrapper.find(".alert-danger").exists()).toBe(true);
    });

    test("Debe de mostrar el navigate al cambiar los parametros", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search"]}>
                <SearchScreen />
            </MemoryRouter>
        );

        wrapper.find("input").simulate("change", {target: {
            name: "searchText",
            value: "batman"
        }});

        wrapper.find("form").prop("onSubmit")({
            preventDefault: () => {}
        });

        expect(mockNavigate).toHaveBeenCalledWith("?q=batman");

    });
    
});