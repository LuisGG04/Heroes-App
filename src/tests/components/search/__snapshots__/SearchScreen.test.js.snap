// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en SearchScreen Debe de mostrar el componente 1`] = `
Array [
  <h1>
    Búsquedas
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h3>
        Buscar
      </h3>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un héroe"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-2"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h3>
        Resultados
      </h3>
      <hr />
      <div
        className="alert alert-info"
      >
        Buscar un héroe
      </div>
    </div>
  </div>,
]
`;
